PLAT-10710: Implement the retry mechanism

symphony-youness opened the pull request:
### Ticket
[PLAT-10710](https://perzoinc.atlassian.net/browse/PLAT-10710)

### Description
Please put here the intent of your pull request.

### Dependencies
List the other pull requests that should be merged before/along this one.

### Checklist
- [x] Referenced a ticket in the PR title and in the corresponding section
- [x] Filled properly the description and dependencies, if any
- [x] Unit tests updated or added
- [x] Docstrings added or updated
- [x] Updated the documentation in [docs folder](../docs)

symphony-youness committed:
PLAT-10710 retry config

symphony-youness committed:
PLAT-10710 Add retry decorator
Retry decorator using a custom AsyncRetrying class
Custom AsyncRetrying class handling asynchronously defined callbacks

symphony-youness committed:
PLAT-10710 Add retry to some services and first try at bot authenticator

symphony-youness committed:
PLAT-10710 Merge with master and add retry to the rest of the services

symphony-youness committed:
PLAT-10710 Fix gen/rest.py modification in previous commit

symphony-youness committed:
PLAT-10710 Authentication Retry and services
Add authentication retry strategy and tests
Add retry strategy to services

symphony-youri commented on examples/authentication/auth.py:
```diff
@@ -7,14 +7,14 @@
 
 async def run():
-    config = BdkConfigLoader.load_from_symphony_dir("config.yaml")
+    config = BdkConfigLoader.load_from_symphony_dir("bot_config/devx3.yaml")
```
To remove from PR maybe (and the commented lines below)

symphony-youri commented on symphony/bdk/core/config/model/bdk_retry_config.py:
```diff
@@ -0,0 +1,61 @@
+        if config is None:
+            config = {}
+
+        self._max_attempts = config.get("maxAttempts") if "maxAttempts" in config else max_attempts
```
would it be interesting to use timedelta to store duration (for the intervals) and avoid wondering what the unit is? (an open question, not a requirement per se)

symphony-elias replied:
IMHO I find that misleading to have the ability to set up `self._max_attempts` from two different places

symphony-youness replied:
Ok, I removed setting `_max_attempts` from the constructor.
It is great for readability to use timedeltas so I'll add that too

symphony-youness replied:
I also removed the properties as we do all the logic in the constructor

symphony-youri commented on symphony/bdk/core/retry/__init__.py:
```diff
@@ -0,0 +1,46 @@
+
+
+def retry(*dargs, **dkw):  # noqa
+    """A decorator that provides a mechanism to repeat requests
```
repeat -> retry (the meaning feels different otherwise)
or mechanism to retry failed requests maybe?

symphony-youri commented on symphony/bdk/core/retry/_asyncio.py:
```diff
@@ -0,0 +1,118 @@
+from tenacity import RetryCallState
+
+
+class AsyncRetrying(BaseRetrying):
```
Maybe there we could comment a bit where is this code coming from why we had to subclass it and even link the PR on tenacity

symphony-youri commented on symphony/bdk/core/retry/startegy.py:
```diff
@@ -0,0 +1,85 @@
+    return isinstance(exception.__cause__, TimeoutError)
+
+
+def is_network_or_minor_or_client_error(exception: Exception) -> bool:
```
is it used?

symphony-youness replied:
It should be used for Datafeed retry, I removed it. Until we do the retry for Datafeed.

symphony-youri commented on symphony/bdk/core/service/presence/presence_service.py:
```diff
@@ -130,6 +140,7 @@ async def read_presence_feed(self, feed_id: str) -> List[V2Presence]:
             feed_id=feed_id)
         return presence_list.value
 
+    @retry(retry=refresh_session_if_unauthorized)
```
don't we want to make this retry parameter the default?

symphony-youness replied:
From what I understood, in java the default one is `isNetworkOrMinorError` so I made the same one and put it as default.

symphony-youness replied:
Sorry, I was wrong on my previous comment. I made use `refresh_session_if_unauthorized` as default.

symphony-youri commented on tests/core/config/__init__.py:
```diff
@@ -0,0 +1,19 @@
+from symphony.bdk.core.config.model.bdk_retry_config import BdkRetryConfig
+
+
+def minimal_retry_config():
```
something worth checking maybe: we should make sure that all tests are still fast to run (default retry configurations were causing some tests to be slow because of the wait between retries for the Java BDK)

symphony-youri replied:
i.e are the config files below with 1s sleep affecting the test executions?

symphony-youness replied:
The configuration files were used only for testing the configuration part (they are not used in other tests)
The methods below returns BdkRetryConfig which should use a wait of 10ms

symphony-youness committed:
PLAT-10710 Remove unused code

symphony-youness committed:
PLAT-10710 Add tenacity to dependencies

symphony-youness committed:
PLAT-10710 Fix config name in auth example

symphony-youness committed:
PLAT-10710 Fix PR comments

symphony-youness committed:
PLAT-10710 Add retry to app and obo authentication
Fix retry tests

symphony-youness commented on symphony/bdk/core/auth/obo_authenticator.py:
```diff
@@ -3,26 +3,16 @@
-        try:
-            return await func(*args, **kwargs)
-        except ApiException as exc:
-            raise AuthUnauthorizedError(OboAuthenticator.unauthorized_message) from exc
```
Instead of raising an AuthUnauthorizedError, I am raising the ApiException (same retry strategy as the authentication)

symphony-youri replied:
With this change are the AuthUnauthorizedError documented below still going to happen?

symphony-youness commented on symphony/bdk/core/auth/ext_app_authenticator.py:
```diff
@@ -50,11 +51,7 @@ async def validate_jwt(self, jwt: str) -> dict:
-        try:
-            pod_certificate = await self._get_pod_certificate()
-        except ApiException as exc:
-            raise AuthInitializationError(f"Unable to get the pod certificate: {exc.reason}") from exc
```
Instead of raising an AuthInitializationError, I am raising the ApiException (same retry strategy as the authentication), we can see from the stack trace that this exception is originating from _get_pod_certificate

symphony-elias commented on examples/authentication/auth.py:
```diff
@@ -12,9 +12,9 @@ async def run():
-        logging.info("Obo example:")
-        obo_auth_session = bdk.obo(username="username")
-        logging.info(await obo_auth_session.session_token)
+        # logging.info("Obo example:")
```
Please remove that from the PR 😉

symphony-elias commented on symphony/bdk/core/auth/obo_authenticator.py:
```diff
@@ -117,10 +109,10 @@ class OboAuthenticatorCert(OboAuthenticator):
+        self._retry_config = retry_config
 
-    @wrap_api_exception
     async def retrieve_obo_session_token_by_user_id(self, user_id: int) -> str:
```
This should be retried, esp. line 124: ` await self._authentication_api.v1_app_user_uid_authenticate_post`

symphony-elias commented on symphony/bdk/core/auth/obo_authenticator.py:
```diff
@@ -133,7 +125,6 @@ async def retrieve_obo_session_token_by_user_id(self, user_id: int) -> str:
         return obo_auth.session_token
 
-    @wrap_api_exception
     async def retrieve_obo_session_token_by_username(self, username: str) -> str:
```
This should be retried as well (line 136)

symphony-elias commented on symphony/bdk/core/config/model/bdk_datafeed_config.py:
```diff
@@ -17,9 +19,11 @@ def __init__(self, config):
         if config is not None:
             self.id_file_path = Path(config.get(DF_ID_FILE_PATH)) if DF_ID_FILE_PATH in config else ""
             self.version = config.get(VERSION)
+            self.retry = BdkRetryConfig(config.get("retry")) if "retry" in config else None
```
You are erasing assignment done in line 22 `if "retry" not in config`

symphony-elias commented on symphony/bdk/core/config/model/bdk_retry_config.py:
```diff
@@ -0,0 +1,61 @@
+    @property
+    def max_attempts(self):
+        if self._max_attempts is None:
+            return self.DEFAULT_MAX_ATTEMPTS
```
IMHO this should be done in the `__init__` method.

symphony-elias commented on symphony/bdk/core/retry/startegy.py:
```diff
@@ -0,0 +1,73 @@
+from tenacity import RetryCallState
```
Please fix the typo: strategy.py instead of sta**r**tegy.py

symphony-elias commented on symphony/bdk/core/retry/startegy.py:
```diff
@@ -0,0 +1,73 @@
+    if isinstance(exception, ApiException):
+        if exception.status >= 500 or exception.status == 401 or exception.status == 429:
+            return True
+    return isinstance(exception.__cause__, TimeoutError)
```
As we discussed, we should also check for the type of `exception`

symphony-elias commented on symphony/bdk/core/retry/startegy.py:
```diff
@@ -0,0 +1,73 @@
+                raise AuthUnauthorizedError(unauthorized_message, exception) from exception
+        raise exception
+    else:
+        return False
```
What happens if `not retry_state.outcome.failed` ? Is this function actually called if outcome is successful ? And you could as well remove the `else` and put `return False` as a by default return value

symphony-youness replied:
Yes, this function is always called. `retry_state.outcome.failed` checks if an exception is being held. I will remove the else statement

symphony-elias commented on symphony/bdk/core/retry/startegy.py:
```diff
@@ -0,0 +1,73 @@
+        exception = retry_state.outcome.exception()
+        if can_authentication_be_retried(exception):
+            return True
+        if isinstance(exception, ApiException):
```
You can merge the two if statements into one: `isinstance(exception, ApiException) and is_unauthorized(exception)` or the check on the type could be put directly in `is_unauthorized`

symphony-elias commented on symphony/bdk/core/service/message/message_service.py:
```diff
@@ -67,6 +74,19 @@ async def send_message(
 
         return await self._messages_api.v4_stream_sid_multi_attachment_message_create_post(**params)
 
+    @retry(retry=refresh_session_if_unauthorized)
```
Why this ?

symphony-youness replied:
removed.

symphony-elias commented on symphony/bdk/core/service/presence/presence_service.py:
```diff
@@ -36,10 +39,12 @@ class OboPresenceService:
+    def __init__(self, presence_api: PresenceApi, auth_session: AuthSession, retry_config: BdkRetryConfig):
         self._presence_api = presence_api
         self._auth_session = auth_session
+        self._retry_config = retry_config
```
BTW how is the retry config passed to `refresh_session_if_unauthorized` ?

symphony-youness replied:
The @retry decorator fetches the `_retry_config` attribute in a specific class to update the retry parameters.
It's done in the retry custom decorator in `retry/__init__.py`

symphony-youness committed:
PLAT-10710 Fix comments
Modified retry config to use timedelta for intervals and removed properties
Added check of ClientConnectorError is retry strategies
Added retry for missed methods in obo_authenticator.py
Fixed typo in strategy.py
Correct auth.py example

symphony-youness committed:
PLAT-10710 Changed default retry and added retry to datafeed

symphony-elias commented on symphony/bdk/core/retry/strategy.py:
```diff
@@ -3,41 +3,70 @@
-def is_unauthorized(exception: ApiException) -> bool:
-    return exception.status == 401
+def is_unauthorized(exception: Exception) -> bool:
+    if isinstance(exception, ApiException):
```
This can even be transformed into a one-liner : `return isinstance(exception, ApiException) and exception.status == 401`.

symphony-elias commented on symphony/bdk/core/retry/strategy.py:
```diff
@@ -3,41 +3,70 @@
     :param exception: The exception to be checked
-    :return:
+    :return: True if the exception is an :py:class:`ApiException` with status 500, 401, 429, 400
+    or is a client timeout error False otherwise
```
Beware the indentation for the second line of the `:return:` part!

symphony-youness replied:
I looked it up on the internet but didn't really get it ? Do you want to add an indent or align it with the previous line (2 indents and a space)

symphony-elias commented on symphony/bdk/core/service/datafeed/abstract_datafeed_loop.py:
```diff
@@ -67,15 +67,16 @@ def __init__(self, datafeed_api: DatafeedApi, auth_session: AuthSession, config:
+        self._running = False
+        self._hard_kill = False
+        self._timeout = None
         self.tasks = []
```
why not prepending `tasks` with `_` ?

symphony-youness committed:
PLAT-10710 Fix comments
missing underscore
indent docstrings
one line return

symphony-youness committed:
PLAT-10710 add retry documentation

symphony-elias commented on symphony/bdk/core/auth/auth_session.py:
```diff
@@ -22,6 +26,7 @@ def __init__(self, authenticator):
     async def refresh(self):
         """Trigger re-authentication to refresh the tokens.
         """
+        logger.debug("Re-authenticate and try again")
```
This could be the first time we log in so this kind of message can be misleading IMHO

symphony-youness replied:
I changed it to "Authenticate", docstring says re-authentication. Maybe we shouldn't use this function when logging for the first time.

symphony-elias commented on docsrc/markdown/datafeed.md:
```diff
@@ -80,4 +80,25 @@ logging.config.dictConfig({
+datafeed:
+  version: 'v2' # specify datafeed version 'v1' or 'v2'
+  retry:
+    maxAttempts: 6 # maximum number of retry attempts
```
Could you document the default values (tell if not provided, it will take the values defined in the root `retry` field ?

symphony-youri replied:
is it going to be infinite by default or the retry defined at root level? (cf bdk_datafeed_config)

symphony-elias replied:
Please explicitly state that if `retry` is not provided in `datafeed`, we are going to have the following default values:
```yaml
retry:
    maxAttempts: -1 # infinite number of attemps
    initialIntervalMillis: 500
    multiplier: 2
    maxIntervalMillis: 300000
```

symphony-elias commented on symphony/bdk/core/retry/__init__.py:
```diff
@@ -0,0 +1,50 @@
+from ._asyncio import AsyncRetrying
+
+
+def retry(*dargs, **dkw):  # noqa
```
You can remove `#noqa`

symphony-elias commented on symphony/bdk/core/retry/__init__.py:
```diff
@@ -0,0 +1,50 @@
+                """Fetches a BdkRetryConfiguration object from the instance of the called function
+
+                If no _retry_config attribute is present, use a default AsyncRetrying configuration
+                Passed retry configuration arguments override the default configuration"""
```
Please add a line break before the trailing `"""`.

symphony-elias commented on symphony/bdk/core/retry/__init__.py:
```diff
@@ -0,0 +1,50 @@
+                If no _retry_config attribute is present, use a default AsyncRetrying configuration
+                Passed retry configuration arguments override the default configuration"""
+                default_kwargs = {}
+                retry_config: BdkRetryConfig = getattr(self, '_retry_config', None)
```
What happens if `retry_config` is None ? I would prefer to have `retry_config: BdkRetryConfig = self._retry_config` and fail fast with an `AttributeError` in case field does not exist.

symphony-youness replied:
I will remove the None default value in gettattr and it should raise an AttributeError if the attribute doesn't exist: 
https://docs.python.org/3/library/functions.html#getattr

symphony-elias commented on symphony/bdk/core/retry/strategy.py:
```diff
@@ -0,0 +1,120 @@
+             False otherwise
+    """
+    if isinstance(exception, ApiException):
+        if exception.status >= 500 or exception.status == 401 or exception.status == 429:
```
instead of `if` + `return`: `return exception.status >= 500 or exception.status == 401 or exception.status == 429`

symphony-elias commented on symphony/bdk/core/retry/strategy.py:
```diff
@@ -0,0 +1,120 @@
+    """
+    if isinstance(exception, ApiException):
+        if exception.status >= 500 or exception.status == 401 or exception.status == 429 or exception.status == 400:
+            return True
```
same here

symphony-youri replied:
also we could reuse the method above here (and just add the check on 400)

symphony-youness replied:
We would be checking twice the isinstance(excpetion, ApiException) is that okay ?
Is this what you meant ?
```python
if isinstance(exception, ApiException):
    return is_network_or_minor_error(exception) or exception.status == 400
 return is_client_timeout_error(exception)
```

symphony-elias commented on symphony/bdk/core/service/session/session_service.py:
```diff
@@ -8,9 +9,11 @@ class OboSessionService:
         self._auth_session = auth_session
+        self._retry_config = retry_config
 
     async def get_session(
```
This should be retried as well.

symphony-elias commented on symphony/bdk/core/service/signal/signal_service.py:
```diff
@@ -25,9 +26,10 @@ class OboSignalService:
         self._auth_session = auth_session
+        self._retry_config = retry_config
 
     async def list_signals(self, skip: int = 0, limit: int = 50) -> SignalList:
```
Methods should be retried here.

symphony-elias commented on symphony/bdk/core/retry/__init__.py:
```diff
@@ -0,0 +1,50 @@
+
+
+def retry(*dargs, **dkw):  # noqa
+    """A decorator that provides a mechanism to to retry failed requests
```
Please document the decorated method should be a class instance method with a valid field `self._retry_config`.

symphony-elias commented on docsrc/markdown/configuration.md:
```diff
@@ -126,3 +137,21 @@ the appId, the private key or certificate for authenticating the extension app.
+
+#### Retry Configuration
+The retry mechanism used by the bot will be configured by these following properties:
+- `maxAttempts`: maximum number of retry attempts that the bot will make.
```
Could you explain how to specify infinite retries (may be useful for datafeed)

symphony-elias commented on tests/core/auth/bot_authenticator_cert_test.py:
```diff
@@ -0,0 +1,81 @@
+
+@pytest.fixture(name="mocked_api_client")
+def fixture_mocked_api_client():
+    def __loader():
```
Please add a comment saying we do this to have different instances of a mocked api client. IMHO `__loader` is not so self-explanatory. `_create_api_client` or similar could be a better name.

symphony-elias commented on tests/core/config/bdk_config_test.py:
```diff
@@ -24,3 +27,43 @@ def test_update_certificate(simple_config_path):
+    config = BdkConfigLoader.load_from_file(config_path)
+
+    assert config.retry.max_attempts == BdkRetryConfig.DEFAULT_MAX_ATTEMPTS
+    assert config.retry.initial_interval.total_seconds() == timedelta(
```
Why testing with `total_seconds()` in each side of the `==` test ? Is that not working with `assert config.retry.initial_interval == timedelta(milliseconds=BdkRetryConfig.DEFAULT_INITIAL_INTERVAL)`?

symphony-youri commented on symphony/bdk/core/config/model/bdk_retry_config.py:
```diff
@@ -0,0 +1,26 @@
+        self.max_attempts = config.get("maxAttempts", self.DEFAULT_MAX_ATTEMPTS)
+        if self.max_attempts < 0:
+            self.max_attempts = sys.maxsize
+        self.initial_interval = config.get("initialIntervalMillis", self.DEFAULT_INITIAL_INTERVAL)
```
that's odd, is that common in Python? to assign a type then another?

symphony-youness replied:
converted it to 1 assignment

symphony-youri commented on symphony/bdk/core/retry/strategy.py:
```diff
@@ -0,0 +1,120 @@
+    return False
+
+
+async def read_datafeed_retry(retry_state: RetryCallState):
```
should this be defined in the abstractdatafeedloop class (if possible) (since it knows the internal of AbstractDatafeedLoop)

symphony-youness replied:
I'd rather group all the strategies in a specific retry module. This would keep tests grouped too.

symphony-youri replied:
ok (we can always refactor later this is internal APIs)

symphony-youness committed:
PLAT-10710 Fix comment remarks

symphony-youness committed:
PLAT-10710 Fix comment remarks
- removed imports and typehints from strategies
- remove bot authenticator > authenticate_bot as it is not used

symphony-youri reviewed (approved):
@symphony-elias your approval matters more than mine :) you followed this closer than I have

symphony-elias commented on symphony/bdk/core/config/model/bdk_config.py:
```diff
@@ -26,10 +27,18 @@ def __init__(self, **config):
+
+        :return: Datafeed retry config if present, Bdk retry config otherwise
+        """
+        return self.retry if self.datafeed.retry is None else self.datafeed.retry
```
`self.datafeed.retry` is never `None` (see constructor of `BdkDatafeedConfig` below). This means we are never going to take the global retry config.
I realize this is the same in the Java BDK, so you could simplify be returning `self.datafeed.retry` directly.

symphony-youness committed:
PLAT-10710 Fix comment remarks
- Remove get_datafeed_retry_config
- Add the description of a default configuration in the documentation

symphony-youness committed:
Merge branch '2.0' into PLAT-10710
